File: backend/app/routers/fantasy.py
"""Fantasy Ping Pong endpoints (feature 007).

A lightweight, password-free identity: the name IS the account. We reuse the same
signed session cookie as the admin area (Starlette ``SessionMiddleware``) but a
separate ``fantasy_user_id`` key, so the two logins are independent. The cookie's
default 14-day lifetime is what "remembered across visits" means here.

Registration, login, the 4-slot team, and CompuBucks all live here. CompuBucks math
is the pure ``app.fantasy`` module (computed on read, never stored).
"""

from __future__ import annotations

from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException, Request, status
from sqlalchemy import select
from sqlalchemy.orm import Session, selectinload

from app.db import get_db
from app.fantasy import compute_compubucks
from app.models import FantasySlot, FantasyUser, Match, Member
from app.schemas import (
    FantasyLoginRequest,
    FantasyRegisterRequest,
    FantasySlotOut,
    FantasyTeamOut,
    FantasyUserOut,
    SessionOut,
    SlotAssignRequest,
)

router = APIRouter(prefix="/api/fantasy", tags=["fantasy"])

SLOT_COUNT = 4


def require_fantasy_user(
    request: Request, db: Session = Depends(get_db)
) -> FantasyUser:
    """Dependency: return the logged-in fantasy user or raise 401.

    Reads ``fantasy_user_id`` from the session cookie. Refuses if there is no
    session or the user no longer exists (e.g. the row was removed).
    """
    user_id = request.session.get("fantasy_user_id")
    if user_id is None:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED, detail="Not logged in"
        )
    user = db.get(FantasyUser, user_id)
    if user is None:
        request.session.pop("fantasy_user_id", None)
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED, detail="Not logged in"
        )
    return user


def _name_key(name: str) -> str:
    """The lookup key for case/space-insensitive identity."""
    return name.strip().lower()


# --- Auth & identity (US1) -------------------------------------------------------

@router.post("/register", response_model=FantasyUserOut, status_code=201)
def register(
    body: FantasyRegisterRequest, request: Request, db: Session = Depends(get_db)
) -> FantasyUserOut:
    """Create a new fantasy account (name + required fun-fact) and log in.

    The schema already trimmed and length-checked the input. Here we only guard the
    name being taken (case/space-insensitive) and set the session.
    """
    key = _name_key(body.name)
    if db.scalar(select(FantasyUser).where(FantasyUser.name_key == key)):
        raise HTTPException(status_code=409, detail="That name is already taken")

    user = FantasyUser(name=body.name, name_key=key, fun_fact=body.fun_fact)
    db.add(user)
    db.commit()
    db.refresh(user)
    request.session["fantasy_user_id"] = user.id
    return FantasyUserOut(id=user.id, name=user.name, fun_fact=user.fun_fact)


@router.post("/login", response_model=FantasyUserOut)
def login(
    body: FantasyLoginRequest, request: Request, db: Session = Depends(get_db)
) -> FantasyUserOut:
    """Log in to an existing account by name only (no password, name is identity)."""
    user = db.scalar(
        select(FantasyUser).where(FantasyUser.name_key == _name_key(body.name))
    )
    if user is None:
        raise HTTPException(status_code=404, detail="No account with that name")
    request.session["fantasy_user_id"] = user.id
    return FantasyUserOut(id=user.id, name=user.name, fun_fact=user.fun_fact)


@router.get("/me", response_model=FantasyUserOut)
def me(user: FantasyUser = Depends(require_fantasy_user)) -> FantasyUserOut:
    """Return the currently logged-in fantasy user (used to decide login vs team)."""
    return FantasyUserOut(id=user.id, name=user.name, fun_fact=user.fun_fact)


@router.post("/logout", response_model=SessionOut)
def logout(request: Request) -> SessionOut:
    """Forget the fantasy user on this device. Safe when already logged out."""
    request.session.pop("fantasy_user_id", None)
    return SessionOut(authenticated=False)


# --- Fantasy team: 4 slots + CompuBucks (US2/US3) --------------------------------

def _validate_slot_index(slot_index: int) -> None:
    if not 1 <= slot_index <= SLOT_COUNT:
        raise HTTPException(
            status_code=422, detail=f"slot_index must be 1-{SLOT_COUNT}"
        )


def _serialize_team(db: Session, user: FantasyUser) -> FantasyTeamOut:
    """Build the always-4-slots view of a user's team plus their CompuBucks total.

    Empty slots come back with null member fields. CompuBucks is computed on read
    from the picked members' completed real games (never stored).
    """
    slots = db.scalars(
        select(FantasySlot)
        .where(FantasySlot.user_id == user.id)
        .options(selectinload(FantasySlot.member).selectinload(Member.team))
    ).all()
    by_index = {s.slot_index: s for s in slots}

    out_slots: list[FantasySlotOut] = []
    for i in range(1, SLOT_COUNT + 1):
        slot = by_index.get(i)
        if slot is None or slot.member is None:
            out_slots.append(
                FantasySlotOut(
                    slot_index=i,
                    member_id=None,
                    member_name=None,
                    team_id=None,
                    team_name=None,
                    team_logo_url=None,
                )
            )
        else:
            m = slot.member
            out_slots.append(
                FantasySlotOut(
                    slot_index=i,
                    member_id=m.id,
                    member_name=m.name,
                    team_id=m.team_id,
                    team_name=m.team.name,
                    team_logo_url=m.team.logo_url,
                )
            )

    matches = db.scalars(
        select(Match).options(selectinload(Match.games))
    ).all()
    total = compute_compubucks(slots, matches)
    return FantasyTeamOut(compubucks=total, slots=out_slots)


@router.get("/team", response_model=FantasyTeamOut)
def get_team(
    user: FantasyUser = Depends(require_fantasy_user), db: Session = Depends(get_db)
) -> FantasyTeamOut:
    """The current user's four slots and their CompuBucks total."""
    return _serialize_team(db, user)


@router.put("/team/slots/{slot_index}", response_model=FantasyTeamOut)
def assign_slot(
    slot_index: int,
    body: SlotAssignRequest,
    user: FantasyUser = Depends(require_fantasy_user),
    db: Session = Depends(get_db),
) -> FantasyTeamOut:
    """Put a real player into a slot (creating or replacing it)."""
    _validate_slot_index(slot_index)

    member = db.get(Member, body.member_id)
    if member is None:
        raise HTTPException(status_code=404, detail="No such player")

    # No duplicate player on the same team (checked in code for a clean message;
    # a UNIQUE constraint also guards it at the DB level).
    existing_same_member = db.scalar(
        select(FantasySlot).where(
            FantasySlot.user_id == user.id,
            FantasySlot.member_id == member.id,
            FantasySlot.slot_index != slot_index,
        )
    )
    if existing_same_member is not None:
        raise HTTPException(
            status_code=409, detail="That player is already on your team"
        )

    slot = db.scalar(
        select(FantasySlot).where(
            FantasySlot.user_id == user.id, FantasySlot.slot_index == slot_index
        )
    )
    if slot is None:
        slot = FantasySlot(
            user_id=user.id, slot_index=slot_index, member_id=member.id
        )
        db.add(slot)
    elif slot.member_id != member.id:
        # A different player — restart this slot's earning clock so the new player
        # only earns from games played after now.
        slot.member_id = member.id
        slot.added_at = datetime.now(timezone.utc).replace(tzinfo=None)
    db.commit()
    return _serialize_team(db, user)


@router.delete("/team/slots/{slot_index}", response_model=FantasyTeamOut)
def clear_slot(
    slot_index: int,
    user: FantasyUser = Depends(require_fantasy_user),
    db: Session = Depends(get_db),
) -> FantasyTeamOut:
    """Empty a slot. Idempotent — clearing an already-empty slot is fine."""
    _validate_slot_index(slot_index)
    slot = db.scalar(
        select(FantasySlot).where(
            FantasySlot.user_id == user.id, FantasySlot.slot_index == slot_index
        )
    )
    if slot is not None:
        db.delete(slot)
        db.commit()
    return _serialize_team(db, user)

File: frontend/src/api/fantasy.ts
// Talks to the Fantasy Ping Pong API (feature 007, see contracts/fantasy-api.md).
// Sends the session cookie with `credentials: 'include'` — that cookie is what
// "remembered across visits" relies on.

// Same base rule as public.ts/admin.ts: empty in production (same-origin via
// Vercel's /api proxy), localhost backend in dev.
const API_BASE = import.meta.env.VITE_API_BASE ?? (import.meta.env.DEV ? 'http://localhost:8000' : '')

// Thrown when the backend rejects a request; carries the status so the UI can
// react (e.g. a 404 on login means "no such name → offer to register").
export class ApiError extends Error {
  status: number
  constructor(status: number, message: string) {
    super(message)
    this.status = status
  }
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, {
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    ...init,
  })
  if (!res.ok) {
    let detail = `Request failed (${res.status})`
    try {
      const body = await res.json()
      if (typeof body.detail === 'string') detail = body.detail
    } catch {
      // no JSON body — keep the generic message
    }
    throw new ApiError(res.status, detail)
  }
  if (res.status === 204) return undefined as T
  return (await res.json()) as T
}

// --- Types ----------------------------------------------------------------------

export interface FantasyUser {
  id: number
  name: string
  fun_fact: string
}

export interface FantasySlot {
  slot_index: number
  member_id: number | null
  member_name: string | null
  team_id: number | null
  team_name: string | null
  team_logo_url: string | null
}

export interface FantasyTeam {
  compubucks: number
  slots: FantasySlot[]
}

export interface Player {
  id: number
  name: string
  team_id: number
  team_name: string
  team_logo_url: string | null
}

// --- Auth & identity (US1) -------------------------------------------------------

export function register(name: string, funFact: string): Promise<FantasyUser> {
  return request<FantasyUser>('/api/fantasy/register', {
    method: 'POST',
    body: JSON.stringify({ name, fun_fact: funFact }),
  })
}

export function login(name: string): Promise<FantasyUser> {
  return request<FantasyUser>('/api/fantasy/login', {
    method: 'POST',
    body: JSON.stringify({ name }),
  })
}

export function getMe(): Promise<FantasyUser> {
  return request<FantasyUser>('/api/fantasy/me')
}

export function logout(): Promise<{ authenticated: boolean }> {
  return request<{ authenticated: boolean }>('/api/fantasy/logout', {
    method: 'POST',
  })
}

// --- Fantasy team (US2) ----------------------------------------------------------

export function fetchTeam(): Promise<FantasyTeam> {
  return request<FantasyTeam>('/api/fantasy/team')
}

export function assignSlot(slotIndex: number, memberId: number): Promise<FantasyTeam> {
  return request<FantasyTeam>(`/api/fantasy/team/slots/${slotIndex}`, {
    method: 'PUT',
    body: JSON.stringify({ member_id: memberId }),
  })
}

export function clearSlot(slotIndex: number): Promise<FantasyTeam> {
  return request<FantasyTeam>(`/api/fantasy/team/slots/${slotIndex}`, {
    method: 'DELETE',
  })
}

export async function fetchMembers(): Promise<Player[]> {
  const data = await request<{ members: Player[] }>('/api/members')
  return data.members
}
